src/rocksdb/src/tuning.rs

```diff
@@ -0,0 +1,461 @@
+    ///
+    /// The default here is the number of cores, as mentioned by
+    /// <https://docs.rs/rocksdb/latest/rocksdb/struct.Options.html#method.increase_parallelism>.
+    ///
+    /// Note that this option is shared across all RocksDB instances that share a `rocksdb::Env`.
+    pub parallelism: Option<i32>,
+
+    /// The most important way to reduce space amplification in RocksDB is compression.
+    ///
+    /// In RocksDB, data on disk is stored in an LSM tree. Because the higher layers (which are
+    /// smaller) will need to be read during reads that aren't cached, we want a relatively
+    /// lightweight compression scheme, choosing `Lz4` as the default, which is considered almost
+    /// always better than `Snappy`.
+    ///
+    /// The meat of the data is stored in the largest, bottom layer, which can be configured
+    /// (using `bottommost_compression_type`) to use a more expensive compression scheme to save
+    /// more space. The default is `Zstd`, which many think has the best compression ratio. Note
+    /// that tuning the bottommost layer separately only makes sense when you have free cpu,
+    /// which we have in the case of the `UPSERT` usecase.
+    pub compression_type: CompressionType,
+
+    /// See `compression_type` for more information.
+    pub bottommost_compression_type: CompressionType,
+}
+
+impl Default for RocksDBTuningParameters {
+    fn default() -> Self {
+        Self {
+            compaction_style: defaults::DEFAULT_COMPACTION_STYLE,
+            optimize_compaction_memtable_budget:
+                defaults::DEFAULT_OPTIMIZE_COMPACTION_MEMTABLE_BUDGET,
+            level_compaction_dynamic_level_bytes:
+                defaults::DEFAULT_LEVEL_COMPACTION_DYNAMIC_LEVEL_BYTES,
+            universal_compaction_target_ratio: defaults::DEFAULT_UNIVERSAL_COMPACTION_RATIO,
+            parallelism: defaults::DEFAULT_PARALLELISM,
+            compression_type: defaults::DEFAULT_COMPRESSION_TYPE,
+            bottommost_compression_type: defaults::DEFAULT_BOTTOMMOST_COMPRESSION_TYPE,
+        }
+    }
+}
+
+impl RocksDBTuningParameters {
+    /// Build a `RocksDBTuningParameters` from strings and values from LD parameters.
+    pub fn from_parameters(
+        compaction_style: &str,
+        optimize_compaction_memtable_budget: usize,
+        level_compaction_dynamic_level_bytes: bool,
+        universal_compaction_target_ratio: i32,
+        parallelism: Option<i32>,
+        compression_type: &str,
+        bottommost_compression_type: &str,
+    ) -> Result<Self, anyhow::Error> {
+        Ok(Self {
+            compaction_style: compaction_style.parse()?,
+            optimize_compaction_memtable_budget,
+            level_compaction_dynamic_level_bytes,
+            universal_compaction_target_ratio: if universal_compaction_target_ratio > 100 {
+                universal_compaction_target_ratio
+            } else {
+                return Err(anyhow::anyhow!(
```

Comment:
untested, might be interesting to see that we handle this correctly, but it's just error handling I guess.